File: pptx-service/renderers/matrix_2x2.py
from builder import add_bg, add_textbox, add_rect, add_line
from theme import ThemeColors


def render(slide, data: dict, t: ThemeColors) -> None:
    add_bg(slide, t.bg)
    heading = data.get("heading", "")
    axes = data.get("axes", {})
    cells = data.get("cells", [{}, {}, {}, {}])
    takeaway = data.get("takeaway", "")

    add_textbox(slide, 0.83, 0.2, 11.0, 0.9, heading,
                t.font_display, 40, t.text, bold=True)

    gx, gy, gw, gh = 1.2, 1.2, 11.0, 5.8

    x_axis = axes.get("x", {})
    y_axis = axes.get("y", {})
    add_textbox(slide, gx, gy + gh + 0.05, gw / 2, 0.35,
                x_axis.get("low", ""), t.font_body, 15, t.muted)
    add_textbox(slide, gx + gw / 2, gy + gh + 0.05, gw / 2, 0.35,
                x_axis.get("high", ""), t.font_body, 15, t.muted, align="right")
    add_textbox(slide, 0.1, gy, 0.9, 1.0,
                y_axis.get("high", ""), t.font_body, 14, t.muted)
    add_textbox(slide, 0.1, gy + gh / 2, 0.9, 1.0,
                y_axis.get("low", ""), t.font_body, 14, t.muted)

    add_line(slide, gx, gy + gh / 2, gx + gw, gy + gh / 2, t.border, 1.0)
    add_line(slide, gx + gw / 2, gy, gx + gw / 2, gy + gh, t.border, 1.0)

    positions = [
        (gx, gy),
        (gx + gw / 2, gy),
        (gx, gy + gh / 2),
        (gx + gw / 2, gy + gh / 2),
    ]
    cell_w = gw / 2 - 0.15
    cell_h = gh / 2 - 0.15

    for i, (cx, cy) in enumerate(positions):
        cell = cells[i] if i < len(cells) else {}
        is_em = cell.get("emphasis", False)
        add_textbox(slide, cx + 0.2, cy + 0.25, cell_w - 0.2, 0.7,
                    cell.get("label", ""), t.font_display, 22,
                    t.accent if is_em else t.text, bold=True, wrap=True)
        desc = cell.get("desc", "")
        if desc:
            add_textbox(slide, cx + 0.2, cy + 1.0, cell_w - 0.2, 1.5,
                        desc, t.font_body, 17, t.muted, wrap=True)

    if takeaway:
        add_textbox(slide, gx, 7.05, gw, 0.4, f"Takeaway: {takeaway}",
                    t.font_body, 18, t.muted, italic=True)

File: lib/pptx/slides/matrix-2x2.ts
import type { Matrix2x2Slide } from '@/lib/types';
import {
  px, py, SLIDE_W,
  setBackground, addRect, addLine, addTextBox, addHeading, addPageNumber,
  type PptxTheme,
} from '../helpers';
import type PptxGenJS from 'pptxgenjs';
type PSlide = ReturnType<InstanceType<typeof PptxGenJS>['addSlide']>;

export function render(slide: PSlide, data: Matrix2x2Slide, t: PptxTheme, n: number, total: number): void {
  setBackground(slide, '#' + t.bg);
  addHeading(slide, data.heading, t, py(60));

  const gridLeft = t.paddingIn + px(30);
  const gridTop = py(200);
  const gridW = SLIDE_W - gridLeft - t.paddingIn;
  const gridH = py(780);
  const midX = gridLeft + gridW / 2;
  const midY = gridTop + gridH / 2;

  addLine(slide, midX, gridTop, midX, gridTop + gridH, '#' + t.rule, 1);
  addLine(slide, gridLeft, midY, gridLeft + gridW, midY, '#' + t.rule, 1);

  const { x: xa, y: ya } = data.axes;
  addTextBox(slide, xa.low, { x: gridLeft, y: gridTop + gridH + py(10), w: gridW / 2, fontSize: t.captionSize, color: '#' + t.muted, fontFace: t.fontBody });
  addTextBox(slide, xa.high, { x: midX, y: gridTop + gridH + py(10), w: gridW / 2, fontSize: t.captionSize, color: '#' + t.muted, fontFace: t.fontBody, align: 'right' });
  addTextBox(slide, ya.high, { x: gridLeft - px(80), y: gridTop, w: px(80), fontSize: t.captionSize, color: '#' + t.muted, fontFace: t.fontBody, align: 'right' });
  addTextBox(slide, ya.low, { x: gridLeft - px(80), y: midY, w: px(80), fontSize: t.captionSize, color: '#' + t.muted, fontFace: t.fontBody, align: 'right' });

  const positions = [
    { x: gridLeft, y: gridTop },
    { x: midX, y: gridTop },
    { x: gridLeft, y: midY },
    { x: midX, y: midY },
  ] as const;
  const cellW = gridW / 2;
  const cellH = gridH / 2;

  data.cells.forEach((cell, i) => {
    const { x, y } = positions[i];
    if (cell.emphasis) {
      addRect(slide, x + px(4), y + py(4), cellW - px(8), cellH - py(8), '#' + t.accent, { transparency: 85 });
    }
    addTextBox(slide, cell.label, { x: x + px(16), y: y + py(16), w: cellW - px(32), fontSize: t.bodySize, color: '#' + t.text, fontFace: t.fontBody, bold: true, wrap: true });
    if (cell.desc) {
      addTextBox(slide, cell.desc, { x: x + px(16), y: y + py(80), w: cellW - px(32), fontSize: t.captionSize, color: '#' + t.muted, fontFace: t.fontBody, wrap: true });
    }
  });

  addPageNumber(slide, n, total, t);
}
